feat: :sparkles: post on publishing Sprout :tada: (#135)
seedcase-project/community: posts/published-sprout/index.qmd

resources](https://sprout.seedcase-project.org/docs/guide/resources/)
that are part of the package.

To create this new data package, you might create a new Python script in
your data package folder called, for example, `main.py`, and then write
the following code:

``` {.python filename="main.py"}
import seedcase_sprout as sp
import polars as pl

properties = sp.PackageProperties(
    # Unique identifier for the package.
    name="diabetes-study",
    title="A Study on Diabetes",
    # You can write Markdown below, with the helper `sp.dedent()`.
    description=sp.dedent("""
        # Data from a 2021 study on diabetes prevalence

        This data package contains data from a study conducted in 2021 on the
        *prevalence* of diabetes in various populations. The data includes:

        - demographic information
        - health metrics
        - survey responses about lifestyle
        """),
    contributors=[
        sp.ContributorProperties(
            title="Luke W. Johnston",
            email="lwjohnst@gmail.com",
            roles=["creator"],
        )
    ]
)
# Save to the `datapackage.json` file in the current folder.
sp.write_properties(properties=properties)
```

> 

Reviewer: Do we want this in `main.py`? I thought we wanted the package properties to be inside the `scripts/properties.py` file?

This code will then create the `datapackage.json` file in the current
folder, while also checking to make sure that the properties are correct
and structured properly.

The most helpful part of Sprout is when you need to add data resources
to the package. Especially when it comes to setting up the metadata. To
help pre-fill in some of the metadata based on the data itself, Sprout

> has the `extract_field_properties()` function. The helpful part comes
> when you want to add data to the package and need a starting point to
> describe the data itself in the metadata. So you can import the data

Reviewer: ```suggestion
has the `extract_field_properties()` function. This function
provides a starting point for the metadata by extracting metadata about the data "fields" (also called columns or variables) from
the data itself. So you can import the data
```

> using Polars and extract the field properties from the data:

Reviewer: ```suggestion
using Polars and extract the field properties from the data like so:
```

``` {.python filename="main.py"}
raw_data = pl.read_csv("path/to/data.csv")
# Extract field properties from the raw, but tidy, data.
field_properties = sp.extract_field_properties(data=raw_data)
```

Which you can then use to insert into a new script with the resource
properties.

``` {.python filename="main.py"}
# Create the resource properties script. This is only created once, it will
# not be overwritten if it already exists.
sp.create_resource_properties_script(
    resource_name="patients",

>     fields=field_properties,

Reviewer: Hm, should we rename the argument to `field_properties`? It's more explicit and correct.

)
```

When you run this code, it will create a new Python script that has the
resource properties listed, along with the pre-filled fields properties

> of the data. Once you've edited this script, you can than use `import`

Reviewer: ```suggestion
of the data. Once you've edited this script, you can then use `import`
```

to import the resource properties into `main.py`.

There's many other things you can do with Sprout, so check out the
[website](https://sprout.seedcase-project.org) for more information,
including how to install it.

## Journey to first release

> Aside from our main goal of having a working package that works as

Reviewer: ```suggestion
Aside from our main goal of having a package that works as
```

designed, we also had a few other goals we wanted to achieve before
publishing our first release of Seedcase Sprout to PyPI:

-   Ensure we meet best practices for formatting, security,
    documentation, and testing.
-   Have an automated release process in place.

> -   Include documentation on the design of Sprout as well as how to use

Reviewer: ```suggestion
-   Include documentation on the design of Sprout as well as a guide on how to use
```

    it.
-   Have a high test coverage (\>90%).
-   Have a comprehensive, documented/explicit build process for local
    development and checks.

When we finally achieved all these goals, that's when we published our
first release to PyPI.

### Challenges

One of the biggest challenges was designing the package, what it would
do, and how it would do it. We made several major design changes
throughout the development of Sprout, that lead to some major rewrites
of the code. But in the end, the final product is something we're quite

> proud of and happy about!

Reviewer: Yay!

After that, the other main challenges were just learning all the

> different things involved in building a well-built, well-tested, secure,

Reviewer: ```suggestion
different things involved in building a well-designed, well-tested, secure,
```

> reliable, and well-documented Python package. One where we can develop

Reviewer: I guess one of the main challenges has also been to set up standardised team workflows following conventions, such as conventional commits, atomic PRs, Agile/Scrum-like project boards and iterations, etc.

Author: I'll add that :grin:

and release continuously with every new change. The fact that we can be
fairly certain that any new pull request that has changes to the code
will be well tested and will automatically release to PyPI is an amazing
feeling, as it simplifies that whole process so much!

### Lessons learned

The most important lessons we learned was that:

-   We should explicitly include a dedicated design brainstorming phase
    at the start of developing a new package. Ideally for a few weeks,
    to give time to brainstorm effectively. And to explicitly try to
    create different designs so that we can compare them and see which
    one is the best fit for the package. We hope that this will mean
    that we do less rewrites later on (though we can't completely avoid

>     that). And to include lots of time to really critique and discuss

Reviewer: ```suggestion
    that since developing and trying out the workflows of the package often uncovers details or sparks ideas that weren’t apparent during the initial brainstorming phase). And to include lots of time to really critique and discuss
```

    the design, so we can be more confident and aligned in the design we
    do eventually choose.

-   An established, smooth, automated, and documented development and
    release process is so helpful as a developer. It means we can really
    focus on the code and text and be confident that changes won't break
    (too many) things. And it means that we don't have to worry or think
    about when or how to release a new version, whether it is secure,
    whether the code in the documentation runs correctly with the new
    changes, and so on.

## Next steps

Our next steps are to take what we learned while building Sprout and

> make a template for creating new data packages as well as new Python

Reviewer: You could add the name of the repo and add a link to it here?

Reviewer: ```suggestion
make a template for creating new data packages ([`template-data-package`](https://github.com/seedcase-project/template-data-package)) as well as new Python
```

packages. That way, for all future Python packages and data packages we

> make, we can use that as a starting point, while also keeping those

Reviewer: ```suggestion
(and others) make, we can use that as a starting point, while also keeping those
```
